add flux-sched container tutorial (#217)
flux-framework/flux-docs: tutorials/containers/flux-sched-container.rst

.. _flux-sched-container:

====================================
Quick Start with Flux in a Container
====================================

Do you want to submit a job to Flux? Here's a short tutorial on how to do so via a container!
The reason this works is because a Flux instance can be run anywhere...

    | You can run it on a cluster... 🥜️
    | You can run it alongside Lustre! 📁️
    | You can run it on a share... 🤗️
    | You can run it anywhere! 🏔️
    | You can run it in a container... 📦️
    | Using Flux? Total no-brainer! 🧠️

Flux has a regularly updated Docker image on `Docker Hub <https://hub.docker.com/u/fluxrm>`_. 
You will need to `install docker <https://docs.docker.com/engine/install/>`_ first.

----------------------
Starting the Container
----------------------

Run the container as follows:

.. code-block:: bash

>   # This says "run an interactive terminal"

Reviewer: in reference to the other comment
```suggestion
  # This says "start an interactive shell under a test Flux instance of size=1"
```

  $ docker run -it fluxrm/flux-sched:latest

> At this point, even before starting an instance, we can inspect what is going on with Flux.

Reviewer: The entrypoint for all the docker images is `/usr/bin/flux start /bin/bash` so a test instance is already started.

Maybe reword to say something like "By default, the docker container starts a test Flux instance of size 1"

.. code-block:: shell

>   # What is the size of resources we have?

Reviewer: ```suggestion
  # What is the size of the current Flux instance (i.e. number of flux-broker processes)
```

  $ flux getattr size
  1

  # Inspect networking

>   $ flux overlay status

Reviewer: This isn't too interesting for an instance of size=1, so it could be dropped.

Author: okay will drop - I'll add this to the docker-compose example since that might be more interesting.

>   # What resources do we have, and what are their statuses?

Reviewer: ```suggestion
  # What resources do we have, and what are their states?
```

  $ flux resource list
       STATE NNODES   NCORES    NGPUS NODELIST
        free      1        4        0 f9004106893d
   allocated      0        0        0 
        down      0        0        0 

In the above, because we are running in a single container, we only see one node with
four cores. Note that you could also imagine an example with docker compose (and we will
be adding this example shortly). Let's look at more Flux interactions, first to see Flux 
environment variables:

.. code-block:: shell

  # Where are different Flux things located (the environment)
  $ flux env

------------
Running Jobs
------------

> And submit a job to get back an id. The "submit" command is headless, and "run" is interactive:

Reviewer: ```suggestion
``flux submit`` submits a job which will be scheduled and run in the background. It prints the jobid to standard output upon successful submission. To run a job interactively, use ``flux run`` which submits and then attaches to the job, displays job output in real time, and does not exit until the job finishes.
```

.. code-block:: shell

    $ flux submit hostname
    ƒM5k8m7m
    $ flux run hostname
    f9004106893d

You can inspect ``--help`` for each of the commands above to see it's possible
to customize the number of nodes, tasks, cores, and other variables.
There are many different options to customize your job submission. For further

> details, please see :core:man1:`flux-submit`.

Reviewer: ```suggestion
details, please see :core:man1:`flux-submit` or run ``flux help submit``.
```

> The identifier shown above is a :ref:`jobid<fluid>` (e.g., ``f9004106893d``). This kind of identifer

Reviewer: ``f9004106893d`` appears to be the hostname. ``ƒM5k8m7m`` is the jobid.

(or similar) is returned for every job submitted, and will be how you interact with your job moving forward. 
Let's throw in a few more sleep jobs, and immediately ask to see them with ``flux jobs``:

.. code-block:: shell

    $ flux submit sleep 360
    $ flux submit sleep 360

.. code-block:: shell

    $ flux jobs
           JOBID USER     NAME       ST NTASKS NNODES     TIME INFO
        ƒq3d755R fluxuser sleep       R      1      1   2.811s 848fc387afd7
        ƒpcByPgK fluxuser sleep       R      1      1   3.805s 848fc387afd7

> But wait, what happened to our first two jobs? Add an ``-a`` for "all" to see all the jobs.

Reviewer: ```suggestion
But wait, what happened to our first two jobs? ``flux jobs`` only shows "active" jobs by default, add an ``-a`` for "all" to see all the jobs.
```

.. code-block:: shell

    $ flux jobs -a
           JOBID USER     NAME       ST NTASKS NNODES     TIME INFO
        ƒq3d755R fluxuser sleep       R      1      1   39.98s 848fc387afd7
        ƒpcByPgK fluxuser sleep       R      1      1   40.97s 848fc387afd7
        ƒDi2hxdm fluxuser hostname   CD      1      1   0.019s 848fc387afd7
        ƒBmTaVZ9 fluxuser hostname   CD      1      1   0.025s 848fc387afd7

.. note::

   See if you can figure out how to list jobs by a particular status, e.g., ``R`` in the output 

>    above means "running." Try ``flux jobs --help``.

Reviewer: ```suggestion
   above means "running." Try ``flux jobs --help`` or ``flux help jobs``.
```

Did you figure it out? It would be ``flux jobs --filter=RUNNING``. What if you were running a long 
process, and you wanted to check on output? Let's do that. Here is script to loop, print, and sleep.

.. code-block:: bash

    #!/bin/bash
    # Save this as loop.sh
    for i in {0..10}; do
        echo "Hello I am loop iteration $i."
        sleep ${i}
    done

Now make the script executable, and submit the job with flux.

.. code-block:: shell

    $ chmod +x ./loop.sh
    $ flux submit ./loop.sh

To see output (and wait until completion) use ``flux job attach``:

.. code-block:: shell

    $ flux job attach ƒ4evXWb9Z
    Hello I am loop iteration 0.
    Hello I am loop iteration 1.
    Hello I am loop iteration 2.
    Hello I am loop iteration 3.
    Hello I am loop iteration 4.
    Hello I am loop iteration 5.

> See :core:man1:`flux-job` for more information on ``flux job attach``.

Reviewer: ```suggestion
See ``flux help job`` or :core:man1:`flux-job` for more information on ``flux job attach``.
```

------------
Viewing Jobs
------------

Aside from listing jobs with ``flux jobs`` there are other ways to get metadata about jobs.

> For your running jobs, you can use ``flux pstree`` to see exactly that - a tree of processes.

Reviewer: ```suggestion
For your running jobs, you can use ``flux pstree`` to see exactly that - a tree of jobs.
```

Let's say we run another sleep job:

.. code-block:: shell

    $ flux submit sleep 350
    ƒ744GwLs

The tree will show us that one sleep!

.. code-block:: shell

    $ flux pstree
    flux
    └── sleep

Submit the same command a few more times? We see that reflected in the tree!

.. code-block:: shell

    $ flux submit sleep 350
    ƒAivupEb
    $ flux submit sleep 350
    ƒB621bj5

    $ flux pstree
    flux
    └── 3*[sleep]

Another way to inspect a job is to look at stats, and pipe the JSON output into jq to make it pretty!

.. code-block:: shell

    $ flux job stats ƒB621bj5  | jq
    {
      "job_states": {
        "depend": 0,
        "priority": 0,
        "sched": 0,
        "run": 3,
        "cleanup": 0,
        "inactive": 0,
        "total": 3
      },
      "successful": 0,
      "failed": 0,
      "canceled": 0,
      "timeout": 0,
      "inactive_purged": 0,
      "queues": []

>     }

Reviewer: `flux job stats` doesn't display information about an individual job, it displays job statistics for the instance as a whole. This is the plumbing equivalent of `flux jobs --stats-only`, so maybe we should drop this section?

Author: okay I can drop! High level dinosaur note  here - as a user I'm always looking for a `flux job <id>` (something) that gives me a json dump of everything I know about a job. I know it's possible with the Python API but I haven't found a good equivalent here. Maybe just `flux job <jobid> --json` or something could do it?

Reviewer: Good idea! I had thought we'd had an issue open on this already, but I cannot find it! The obvious choice might be `flux job info --all JOBID`, which would dump all known info "keys" in a single JSON object.
